swiftsimio/initial_conditions/generate_particles.py

```diff
@@ -1,5 +1,1047 @@
-"""
-Particle generation code.
+"""Generate SPH initial conditions for SPH simulations iteratively for a given density function following Arth et al.
+2019 (https://arxiv.org/abs/1907.11250).
 
-TBD
 """
+
+# --------------------------------------------
+# IC generation related stuff
+# --------------------------------------------
+
+
+import numpy as np
+import unyt
+from unyt import unyt_array
+from math import erf
+from typing import Union
+
+from .IC_kernel import get_kernel_data
+from swiftsimio.optional_packages import KDTree, TREE_AVAILABLE
+from swiftsimio import Writer
+
+
+seed_set = False
+
+
+class ic_sim_params(object):
+    r"""
+    object containing simulation parameters
+    """
+    boxsize: unyt_array = unyt_array([1.0, 1.0, 1.0], "cm")
+    periodic: bool = True
+    nx: int = 100
+    ndim: int = 2
+    unit_l: Union[None, unyt.unit_object.Unit, str] = None
+    unit_m: Union[unyt.unit_object.Unit, str] = "g"
+    kernel: str = "cubic spline"
+    eta: float = 1.2348
+
+    def __init__(
+        self,
+        boxsize: unyt_array = unyt_array([1.0, 1.0, 1.0], "cm"),
+        periodic: bool = True,
+        nx: int = 100,
+        ndim: int = 2,
+        unit_l: Union[None, unyt.unit_object.Unit, str] = None,
+        unit_m: Union[unyt.unit_object.Unit, str] = "g",
+        kernel: str = "cubic spline",
+        eta: float = 1.2348,
+    ):
+        r"""
+        Set up the simulation parameters for the initial conditions you want to
+        generate. The dict it returns is a necessary argument to call
+        ``generate_IC_for_given_density()``
+
+
+        Parameters
```

Comment:
You can then move this up to the top-level docsrting under `attributes`

Reply:
'tis done.